orttraining/orttraining/python/training/_ortmodule_io.py

```diff
@@ -55,12 +64,19 @@ def _convert_input_to_list(param_names, user_input_names, buffer_names, inputs,
     non_none_inputs = [inp for inp in inputs if inp is not None]
     named_buffers_iter = iter(buffer_names)
     result = []
-    for input_idx, name in enumerate(user_input_names):
+
+    for input_idx, name in enumerate(onnx_input_names):
         inp = None
-        if input_idx < len(non_none_inputs):
-            inp = non_none_inputs[input_idx]
-        elif name in kwargs and kwargs[name] is not None:
+        if name in kwargs and kwargs[name] is not None:
+            # Only use keywords coming from user that are expected by ONNX model
             inp = kwargs[name]
+        elif input_idx < len(non_none_inputs):
+            # Only use positionals coming from user that are expected by ONNX model
+            if name != input_info.names[input_idx]:
+                # When ONNX drops unused inputs, get correct index from user input
+                input_idx = input_info.names.index(name)
```

Comment:
Just checking: will we only search `input_info.names` for the names of arguments (and hence expect them to be in the first `len(non_none_inputs)` elements of the list)?

If so, linear search with `index()` is probably fine. If not, and we might be searching through hundreds or thousands of strings, I'd suggest adding a dictionary-based map.

Reply:
Yes. `input_info.names` only contains input names from original pytorch `forward(self, ...)` method. 
It doesn't include initializer names nor pytorch registered buffers or anything else. Should be just a few elements long